src/humancellatlas/data/metadata/helpers/staging_area.py

```diff
@@ -0,0 +1,323 @@
+                entity_id = link['process_id']
+                entity_ids[entity_type].add(entity_id)
+                for category in ('input', 'output', 'protocol'):
+                    for file in link[f'{category}s']:
+                        entity_type = file[f'{category}_type']
+                        entity_id = file[f'{category}_id']
+                        entity_ids[entity_type].add(entity_id)
+            elif link_type == 'supplementary_file_link':
+                for file in link['files']:
+                    entity_type = file['file_type']
+                    entity_id = file['file_id']
+                    entity_ids[entity_type].add(entity_id)
+            else:
+                raise ValueError('Unknown link type', link_type)
+        return dict(entity_ids)
+
+    def get_bundle_metadata(self, subgraph_id: str) -> Tuple[str, List[JSON], JSON]:
+        """
+        Return a tuple consisting of the version of the downloaded bundle, a
+        list of the manifest entries for all metadata files in the bundle, and a
+        dictionary mapping the file name of each metadata file in the bundle to
+        the JSON contents of that file.
+        """
+        logger.debug('Composing bundle %s', subgraph_id)
+
+        links_file = self.links[subgraph_id]
+        manifest = []
+        metadata = {
+            'links.json': links_file.content
+        }
+        entity_ids_by_type = self.entity_ids_by_type(subgraph_id)
+        for entity_type, entity_ids in entity_ids_by_type.items():
+            # Sort entity_ids to produce the same ordering on multiple runs
+            for i, entity_id in enumerate(sorted(entity_ids)):
+                json_file_name = f'{entity_type}_{i}.json'
+                metadata_file = self.metadata[entity_id]
+                json_content = metadata_file.content
+                metadata[json_file_name] = json_content
+                file_manifest = {
+                    'content-type': 'application/json;',
+                    'crc32c': '0' * 8,
+                    'indexed': True,
+                    'name': json_file_name,
+                    's3_etag': None,
+                    'sha1': None,
+                    'sha256': '0' * 64,
+                    'size': len(json.dumps(json_content)),
+                    'uuid': metadata_file.uuid,
+                    'version': metadata_file.version
+                }
+                manifest.append(file_manifest)
+                if entity_type.endswith('_file'):
+                    file_manifest = self.descriptors[entity_id].to_manifest_entry
+                    manifest.append(file_manifest)
+                else:
+                    pass
+        logger.debug('Bundle contents: %s', list(metadata.keys()))
+        return links_file.version, manifest, metadata
+
+
+@attr.s(frozen=True, kw_only=True, auto_attribs=True)
+class GitHubStagingAreaFactory:
+    repo: Repository
+    branch: str
```

Comment:
This can be a branch, a sha or a tag so

```suggestion
    ref: str
```

https://git-scm.com/book/en/v2/Git-Internals-Git-References